### PR TITLE
Use predicate function in `lstrip` etc.

### LINKED ISSUE #27211: unicode whitespace not recognized by lstrip()
Consider:

```
julia> a = " \U02009 foo"
"   foo"

julia> lstrip(a)
"  foo"
```

Note that `\U02009` is a `Unicode Character 'THIN SPACE'` character.

I don't know much about string processing or unicode, but it seems to me that `lstrip()` is doing something that's perhaps too naive:

```
const _default_delims = [' ','\t','\n','\v','\f','\r']
...
function lstrip(s::AbstractString, chars::Chars=_default_delims)
```

I realize you can override `chars` but I would suggest that we expand the default character set for string trimming to include unicode separators (ref https://www.fileformat.info/info/unicode/category/Zs/list.htm).

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #27211. Should be non-breaking.

### CODE REVIEW

File: base/strings/util.jl [hunk 1]
```diff
@@ -133,22 +140,32 @@ julia> lstrip(a)
...
+function lstrip(f, s::AbstractString)
     e = lastindex(s)
     for (i, c) in pairs(s)
-        !(c in chars) && return SubString(s, i, e)
+        !f(c) && return SubString(s, i, e)
     end
     SubString(s, e+1, e)
 end
+lstrip(s::AbstractString) = lstrip(isspace, f)
+lstrip(s::AbstractString, chars::Chars) = lstrip(c -> c in chars, s)
```
Comment (JeffBezanson) [inline base/strings/util.jl]: `in(chars)`

**Comment by fredrikekre** [comment]
Perhaps the asymmetry between `strip(this_decides_what_to_strip, str)` (`strip(f, str)`) and  `strip(str, this_decides_what_to_strip)` (`strip(str, chars)`) is a bit unfortunate? Can we deprecate the latter in favor of the former?

**Comment by simonbyrne** [comment]
I'll leave the decision on the `chars` argument to triage.

File: base/strings/util.jl [hunk 2]
```diff
@@ -159,28 +176,44 @@ julia> rstrip(a)
...
 """
-function rstrip(s::AbstractString, chars::Chars=_default_delims)
+function rstrip(f, s::AbstractString)
     for (i, c) in Iterators.reverse(pairs(s))
-        c in chars || return SubString(s, 1, i)
+        f(c) || return SubString(s, 1, i)
     end
     SubString(s, 1, 0)
 end
+rstrip(s::AbstractString) = rstrip(isspace, f)
```
Comment (KDr2) [inline base/strings/util.jl]: A typo? 
`rstrip(s::AbstractString) = rstrip(isspace, f)` should be `rstrip(s::AbstractString) = rstrip(isspace, s)`?

Comment (KDr2) [reply base/strings/util.jl]: Seems `lstrip`  also has this typo.

**Comment by simonbyrne** [comment]
Should we deprecate `chars`?

**Comment by simonbyrne** [comment]
Ah, I assume that is what you were implying.

**Comment by simonbyrne** [comment]
Note that `split`/`rsplit` uses `split(string, predicate)`.

**Comment by ararslan** [comment]
Looks like there's a still a use of the deprecated `rstrip` method. This should do it:
```diff
--- a/base/mpfr.jl
+++ b/base/mpfr.jl
@@ -932,7 +932,7 @@ function _prettify_bigfloat(s::String)::String
     if !occursin('.', mantissa)
         mantissa = string(mantissa, '.')
     end
-    mantissa = rstrip(mantissa, '0')
+    mantissa = rstrip(==('0'), mantissa)
     if endswith(mantissa, '.')
         mantissa = string(mantissa, '0')
     end
```

**Comment by ararslan** [comment]
Another one:
```diff
--- a/stdlib/Markdown/src/GitHub/GitHub.jl
+++ b/stdlib/Markdown/src/GitHub/GitHub.jl
@@ -9,7 +9,7 @@ function fencedcode(stream::IO, block::MD)
         skip(stream, -1)
         ch = read(stream, Char)
         trailing = strip(readline(stream))
-        flavor = lstrip(trailing, ch)
+        flavor = lstrip(==(ch), trailing)
         n = 3 + length(trailing) - length(flavor)
 
         # inline code block
```

**Comment by simonbyrne** [comment]
Can you point to the broken docstrings? I can't see any.

**Comment by simonbyrne** [comment]
Looking at https://github.com/JuliaDocs/Documenter.jl/pull/731, I'm not 100% convinced it's a strict improvement.

Additionally, considering that the behaviour will be different than `split`, I'm wondering if we should switch back.

**Comment by ararslan** [comment]
> Can you point to the broken docstrings? I can't see any.

They're showing up in the Travis log: https://travis-ci.org/JuliaLang/julia/jobs/383852269#L2958.

> Additionally, considering that the behaviour will be different than `split`, I'm wondering if we should switch back.

Does there need to be a 1-1 correspondence with `split`? They're related functions but not _that_ related.

**Comment by nalimilan** [comment]
I agree it sounds a bit weird to require using `==` or `isequal` here. These are string-specific functions, so it makes sense to support more succinct syntaxes for stripping a character or substring. (Actually we could even allow this syntax with `find*` functions in 1.x, I didn't propose them because the priority was to get the generic API right.)

**Comment by StefanKarpinski** [comment]
I have to agree that there doesn't seem to be much harm in leaving the old `strip` signatures. Having the character, character set or regex to strip as an optional second argument makes sense; having a predicate as a first argument also makes sense. As long as the argument to be stripped must be an `AbstractString` (which seems sensible), having a little variation in argument order doesn't seem like the worst thing.

**Comment by nalimilan** [comment]
The current `strip(string, chars)` sounds fine. I'm not sure why, but I find `strip(chars, string)` unnatural; probably because `strip` transforms `string` and `chars` is only a secondary argument (which is indeed optional).

**Comment by StefanKarpinski** [comment]
I also don't think we should switch the order. Is there much to do here then?

**Comment by simonbyrne** [comment]
I think we should still allow predicate functions, since they can be much more efficient (e.g. checking ranges, short-circuiting if there are no non-ASCII chars).

I've opened a simpler PR: #27309

**Comment by simonbyrne** [comment]
Closed in favour of #27309 (for now...)

**Pull request closed without merging**